content/en/blog/_posts/2021-11-18-prevent-persistentvolume-leaks-when-deleting-out-of-order.md

```diff
@@ -0,0 +1,199 @@
+```
+NAME                        STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS               AGE
+example-vanilla-block-pvc   Bound    pvc-6791fdd4-5fad-438e-a7fb-16410363e3da   5Gi        RWO            example-vanilla-block-sc   19s
+```
+
+#### Delete PV
+
+When I try to delete a bound PV, the cluster blocks and the `kubectl` tool does
+not return back control to the shell; for example:
+
+```
+kubectl delete pv pvc-6791fdd4-5fad-438e-a7fb-16410363e3da
+```
+
+```
+persistentvolume "pvc-6791fdd4-5fad-438e-a7fb-16410363e3da" deleted
+^C
+```
+
+Retrieving the PV:
+```
+kubectl get pv pvc-6791fdd4-5fad-438e-a7fb-16410363e3da
+```
+
+It can be observed that the PV is in `Terminating` state
+```
+NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS        CLAIM                               STORAGECLASS               REASON   AGE
+pvc-6791fdd4-5fad-438e-a7fb-16410363e3da   5Gi        RWO            Delete           Terminating   default/example-vanilla-block-pvc   example-vanilla-block-sc            2m23s
+```
+
+#### Delete PVC
+
+```
+kubectl delete pvc example-vanilla-block-pvc
+```
+
+The following output is seen if the PVC gets successfully deleted:
+```
+persistentvolumeclaim "example-vanilla-block-pvc" deleted
+```
+
+The PV object from the cluster also gets deleted. When attempting to retrieve the PV
+it will be observed that the PV is no longer found:
+
+```
+kubectl get pv pvc-6791fdd4-5fad-438e-a7fb-16410363e3da
+```
+
+```
+Error from server (NotFound): persistentvolumes "pvc-6791fdd4-5fad-438e-a7fb-16410363e3da" not found
+```
+
+Although the PV is deleted the underlying storage resource is not deleted, and
+needs to be removed manually.
+
+To sum it up, the reclaim policy associated with the Persistent Volume is currently
+ignored under certain circumstance. For a `Bound` PV-PVC pair the ordering of PV-PVC
+deletion determines whether the PV reclaim policy is honored. The reclaim policy 
+is honored if the PVC is deleted first, however, if the PV is deleted prior to 
+deleting the PVC then the reclaim policy is not exercised. As a result of this behavior,
+the associated storage asset in the external infrastructure is not removed.
+
+## PV reclaim policy with Kubernetes v1.23
+
+The new behavior ensures that the underlying storage object is deleted from the backend when users attempt to delete a PV manually.
+
+#### How to enable new behavior?
+
+To make use of the new behavior, you must have upgraded your cluster to the v1.23 release of Kubernetes.
+You need to make sure that you are running the CSI [`external-provisioner`](https://github.com/kubernetes-csi/external-provisioner) version `4.0.0`, or later.
+You must also enable the `HonorPVReclaimPolicy` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) for the
+`external-provisioner` and for the `kube-controller-manager`.
```

Comment:
This feature gate is for `kube-controller-manager`.  There is a way to specify that when starting `kube-controller-manager`.  

`external-provisioner` just checks this same feature gate. User does not specify this feature gate when starting `external-provisioner` .  Please clarify that.

Reply:
Thanks for spotting this @xing-yang.

> `external-provisioner` just checks this same feature gate.

How's that done? (I didn't think an API client has a way to check what feature gates are enabled for control plane components). Maybe the answer is that you need to turn on the feature gate for the controller manager and then so long as you're using v4.0.0 or later of external-provisioner, you get the behavior you want anyway?

I think I'm therefore also asking: how does the `external-provisioner` decide whether or not to set the finalizer `external-provisioner.volume.kubernetes.io/finalizer` ?
(If the answer's complicated we don't need to mention it in this article)

Reply:
Just chatted with @deepakkinni.  We actually do need to specify the same feature gate when starting the `external-provisioner`.  Please ignore my comments.